### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by stoffern**
Need to complete the "Quick Start" and "API" section.
I can merge it when it is complete if that is ok? :wink: 

closes: https://github.com/nodkz/graphql-compose/issues/6

### CODE REVIEW

**Comment by nodkz** [comment]
Let I bring some clarifications about `graphql-compose` and its plugins.

### graphql-compose - type manipulator, schema builder
- allow easily edit `GraphQL` types (add/remove fields)
- provide an easy way for creating relations between types via `Resolver`s (one time describe outputType, args and resolve method and then use them as you wish). 
- provide converter from `OutputType` to `InputType`
- provide `projection` parser from AST

lack in tests, due changing API in past

### [graphql-compose-connection](https://github.com/nodkz/graphql-compose-connection) - resolver constructor (TypeTC -> TypeTC + 1 resolver)
- It is `connection` resolver constructor, that use `findMany` and `count`resolvers which creates for you [RelayConnectionType](https://facebook.github.io/relay/graphql/connections.htm).
- adds `sort` and `filter` to args (no one library does not realize this complex logic. It store `sort` and `filter` values in `cursor`). How it works described [here](https://github.com/RisingStack/graffiti-mongoose/issues/99#issuecomment-234524046)

Well tested with mocha and chai.

### [graphql-compose-mongoose](https://github.com/nodkz/graphql-compose-mongoose) - model converter (MongooseModel -> TypeTC)
- It's mongoose model converter, that using `graphql-compose` for building one main type ObjectType (`graphql-compose` helps to convert ObjectType to InputObject)
- Create commonly used `Resolvers` with a bunch of so-called subtypes (filter, args, sorts, inputs). For providing build relations in your schema with `graphql-compose`.
- Internally used `graphql-compose-connection` for creating `connection` resolvers. Also it checks indexes on your mongoose fields and adds them to `sort` arg.

Well tested with mocha and chai.

### [graphql-compose-relay](https://github.com/nodkz/graphql-compose-relay) - type converter (TypeTC -> TypeTC prepared for Relay)
- adds Node interface to wrapped Type
- adds `id` field or wrap existed in MainType, for providing global uniq id via base64(typeName:id)
- adds `node` field to `rootQuery` that resolves objects by it's global id
- traverse by `mutation resolvers` and change its outputTypes adds `clientMutationId`, wraps `resolver` to pass `clientMutationId` from args to the `OutputType`
- traverse by `all resolvers` and move all args into one `input` arg, for Relay compatibility  

Well tested with mocha and chai.

### Immediate goal
- Obtain better filtering customization (https://github.com/nodkz/graphql-compose/issues/3)

### In future
- Add [DataLoader](https://github.com/facebook/dataloader) wrapper for `findById` resolvers.
- Some better API for building schema with restrictions in `graphql-compose`
- Somehow implement `subscriptions`

PS. 3 days later, I'm going on vacation for 2-3 weeks. So I'll try be online via mobile, to answer on easy questions but can not provide answers with code examples. So if you have hard questions, feel free to open issues in this couple of days (from Friday I will be via mobile).

**Comment by stoffern** [comment]
@nodkz dont you think this text about the plugins should be put into the respective plugins?
when there are 20 plugins the page will be quite long with all this text :wink:

**Comment by nodkz** [comment]
I think for the first time it should be. When number of plugins will be more we always can remove its descriptions.

**Comment by nodkz** [comment]
Also in 0.0.24 I fixed projection option for fieldConfig.

Like `addRelation` has `projection`:
```js
AddressTC.addRelation(
  'country',
  () => ({
    resolver: CountryTC.getResolver('findById').wrapResolve(next => resolveParams => {
       return resolveParams.source.country_id ? next(resolveParams) : null,
    }),
    args: {
       _id: (source) => source.country_id,
    },
    projection: { country_id: true }, // if requested `country` fields, also request `country_id` from Address document
  })
);
```

Same way you may create additional fields to type with `projection` option (to standard GraphQL field config I added `projection` param):
```js
UserSchema = mongoose.Schema({
  firstName: String,
  lastName: String,
});

UserTC.addField(
  'fullName',
  {
    type: GraphQLString,
    resolver: (source) => `${source.firstName} ${source.lastName}`,
    projection: { firstName: true, lastName: true }, 
  })
);
```
Now with GraphQL query
```js
{
  user {
    fullName
  }
}
```

You will get `UserName UserLastname`. If you do not provide `projection` result will be `undefined undefined`.

**Comment by nodkz** [comment]
1) About `ResolverWiddleware`: 
I think to **remove** them at all. They have complex API. And I never used them in my app.

2) About code coverage: 
I was pleasantly surprised that all plugins have a code coverage more than 90%. `graphql-compose` has only 40% (right now it well tested in our private app). I can write tests only in the next month. So I think not show a coverage badge until it becomes not less 70-80 percents.

3) Also forgot to point, that `graphql-compose-mongoose` support a `mongoose.Schema.Types.Mixed` type with input args and output types (you may get or set any JSON object to field/arg). Examples [with query](https://graphql-compose.herokuapp.com/user/?query=%0A%7B%0A%20%20userById(_id%3A%20%2257bb44dd21d2befb7ca3f001%22)%20%7B%0A%20%20%20%20_id%0A%20%20%20%20name%0A%20%20%20%20someMixed%0A%20%20%7D%0A%7D%0A%20%20%20%20%20%20), [with mutation](https://graphql-compose.herokuapp.com/user/?query=%0Amutation%20%7B%0A%20%20userCreate(record%3A%20%7B%0A%20%20%20%20name%3A%20%22My%20Name%22%2C%0A%20%20%20%20age%3A%2024%2C%0A%20%20%20%20gender%3A%20ladyboy%2C%0A%20%20%20%20contacts%3A%20%7B%0A%20%20%20%20%20%20email%3A%20%22mail%40example.com%22%2C%0A%20%20%20%20%20%20phones%3A%20%5B%0A%20%20%20%20%20%20%20%20%22111-222-333-444%22%2C%0A%20%20%20%20%20%20%20%20%22444-555-666-777%22%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%7D%2C%0A%20%20%20%20someMixed%3A%20%7B%0A%20%20%20%20%20%20a%3A%201%2C%0A%20%20%20%20%20%20b%3A%202%2C%0A%20%20%20%20%20%20c%3A%20%5B%201%2C%202%2C%203%2C%20true%2C%20false%2C%20%7B%20sub%3A%201%20%7D%5D%0A%20%20%20%20%7D%0A%20%20%7D)%20%7B%0A%20%20%20%20recordId%0A%20%20%20%20record%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20age%0A%20%20%20%20%20%20gender%0A%20%20%20%20%20%20contacts%20%7B%0A%20%20%20%20%20%20%20%20email%0A%20%20%20%20%20%20%20%20phones%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20someMixed%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%20%20%20%20%20%20)

4) `graphql-compose-mongoose` loves `indexes` on fields or `compound indexes` (with several fields). If it founds indexed fields, it automaticaly adds them to `sort` arg in resolvers. It's better than add fields explicitly via options of `composeWithMongoose(Model, opts)`.

5) About naming of imports, I think better namimg is:
```js
import composeWithMongoose from 'graphql-compose-mongoose';
import composeWithRelay from 'graphql-compose-relay';
import composeWithConnection from 'graphql-compose-connection';
```

6) GraphQL-compose hoisting problem friendly and based on practical experience of the construction of complex schemas. 

I recommend defining `mongoose schemas`/`models` separately, one per file. Do not mix several models in one file. Define `TypeComposer`, right after mongoose model definitions in same file.
Eg.
```js
export const CountrySchema = new mongoose.Schema({
  code: String,
  // ...
});
export const Country = mongoose.model('Country', CountrySchema);

export const CountryTC = composeWithMongoose(Country);
// ... some other stuff for defining graphql type via TypeComposer, eg. relations, restrictions
```
The main idea that mongoose model and graphql type should be described fully in one place/file and provides clear import in other files. Eg `import { CountrySchema, Country, CountryTC } from './country';`. You always may obtain mongoose schema, model, or type composer. Also this rule helps to modify your code in future (eg. found some type via GraphiQL, so you clearly understand that all this type defined in one particular file. I had [bad experience](https://github.com/nodkz/unpublished-disgusting-graphql-mongoose-module) when define type in several files).
More examples of defining schemas with relations can be found [here](https://github.com/nodkz/graphql-compose-examples/tree/master/examples/northwind) in folder `models`.

By this convention, constructed `addRelation` method which uses arrow function. This is done on purpose to solve hoisting problem (when 2 models import each other). Without arrow function we get empty value when importing `ModelTC`. Of cource I can simplify this method (add variant without arrow function), but when in future you will extend ModelTC, you may run up on cross-referenced models (directly, or via intermediate model files) and it error is difficult to diagnose. 

PS.  If I have something to remember, then I'll write with my "outstanding" English 😝

**Comment by nodkz** [comment]
I'll try to find time at the end of the week, review and merge your changes.
If you have some additionals, please add them.
Thanks.

PS. I'm back from vacation. Right now I focused on our internal product changes but definitely find time for OSS.

**Comment by stoffern** [comment]
👍 great to have you back. hope you had a great vacation 😉 
I have some more additions, in both API and the Quick Start guide.

I will try to complete them tomorrow.

**Comment by nodkz** [comment]
Awesome! Will wait. 👍

PS. 2 weeks, 6 flights, 1700km by car. It was very active vacation 😉 
And I so happy to come back to my workplace 😂

**Comment by stoffern** [comment]
@nodkz i think that is it.. 
There is alot of improvements that can be done, but there will always be that 😉 

Please look throu it and see if you have any adjustments, i guess it could be a good idea to publish them after that and let others and us update them continually

**Comment by nodkz** [comment]
Sorry, but I delay accepting this PR.

On the previous week, I improve TypeComposer: https://github.com/nodkz/graphql-compose/issues/22.
Today improve type creation: [v0.2.0](https://github.com/nodkz/graphql-compose/releases/tag/v0.2.0)

And right now I want to make breaking changes in the `Resolver`: https://github.com/nodkz/graphql-compose/issues/23.

**Comment by stoffern** [comment]
I will update the docs for the lastest update.
Have alot of things to do right now and are away for 1 week.

Will look into it when i get back, but feel free to look into the docs in the meanwhile 😉

**Comment by nodkz** [comment]
Firstly I should integrate searching in our internal project. And test how it is to work with current API.
Right now I found some features (api sugar) that definitely should be added before publishing docs (cause it reduce app code).

**Comment by stoffern** [comment]
Sorry for the delay here, my pockets are completely full atm.
I still have this in mind, but i just need more time 😉

**Comment by nodkz** [comment]
Don't worry! Same things. 😉

File: docs/03-guide/04-relations-and-connections.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'People'
+  }],
+})
+export const Companies = mongoose.model('Companies', CompanysSchema)
+export const CompaniesTC = composeWithMongoose(Companies)
+```
+
+###Adding a single relation
+Lets start of building a connection from the people to theyr jobs.
```
Comment (ikari-pl) [inline docs/03-guide/04-relations-and-connections.md]: their* 
(not a grammar nazi comment! Just noticing a typo :D)

File: docs/03-guide/07-relay.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+###Add support in TypeComposers
+It cant be more simple than this:
+```js
+export const PersonTC = composeWithRelay(composeMongoose(Person))
+```
+
+###Add support in RootQuery
+```js
+composeFromRelay(GQC.rootQuery())
```
Comment (ikari-pl) [inline docs/03-guide/07-relay.md]: Ambiguity here: is this **in addition** to `composeWithRelay` from example above, or **instead**?

**Comment by ikari-pl** [comment]
Guys, this PR is extremely useful in comparison to current docs. As a user, I'd suggest merging it and doing all necessary improvements later.

**Comment by nodkz** [comment]
@ikari-pl indeed, let move it

**Comment by nodkz** [comment]
Couple minutes ago finished to review and actualize current documentations.
@stoffern @ikari-pl  please review it if u have couple minutes.

@stoffern thanks a lot for awesome work 👍👍👍

**Pull request merged**